src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java

```diff
@@ -154,12 +171,130 @@ protected AuthResult authenticateUser(final String username, final String passwo
+                    .withApiErrors(DefaultApiError.INTERNAL_SERVER_ERROR)
+                    .withExceptionCause(e)
+                    .withExceptionMessage("Error parsing the embedded auth data from Okta.")
+                    .build();
+        }
+    }
+
+    /**
+     * Get list of enrolled MFA factors for a user from their authentication result
+     * @param authResult  Okta response object after successful authentication
+     * @return List of factors
+     */
+    protected List<Factor> getUserFactorsFromAuthResult(final AuthResult authResult) {
+
+        final EmbeddedAuthResponseDataV1 embeddedAuthData = this.getEmbeddedAuthData(authResult);
+
+        if (embeddedAuthData != null && embeddedAuthData.getFactors() != null) {
+            return embeddedAuthData.getFactors();
+        } else {
+
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.INTERNAL_SERVER_ERROR)
+                    .withExceptionMessage("Could not parse user factors from Okta response.")
+                    .build();
+        }
+    }
+
+    /**
+     * Get user id from authentication result
+     * @param authResult Okta response object after successful authentication
+     * @return The u
+     * ser ID
+     */
+    protected String getUserIdFromAuthResult(final AuthResult authResult) {
+        final EmbeddedAuthResponseDataV1 embeddedAuthData = this.getEmbeddedAuthData(authResult);
+
+        if (embeddedAuthData.getUser() != null) {
+            return embeddedAuthData.getUser().getId();
+        } else {
+
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.INTERNAL_SERVER_ERROR)
+                    .withExceptionMessage("Could not parse user data from Okta response.")
+                    .build();
+        }
+    }
+
+    /**
+     * Get username from authentication result
+     * @param authResult Okta response object after successful authentication
+     * @return The username
+     */
+    String getUserLoginFromAuthResult(final AuthResult authResult) {
+        final EmbeddedAuthResponseDataV1 embeddedAuthData = this.getEmbeddedAuthData(authResult);
+
+        try {
+            return embeddedAuthData.getUser().getProfile().getLogin();
+        } catch(NullPointerException npe) {
+
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.INTERNAL_SERVER_ERROR)
+                    .withExceptionMessage("Could not parse user data from Okta response.")
+                    .build();
+        }
+    }
+
+    /**
+     * Ensure the user has at least one active MFA device set up.
+     * @param factors - List of user factors
+     */
+    protected void validateUserFactors(final List<Factor> factors) {
+
+       if(factors == null || factors.isEmpty() || factors.stream()
+               .allMatch(factor -> StringUtils.equals(factor.getStatus(), MFA_FACTOR_NOT_SETUP_STATUS)))
```

Comment:
I agree, but Okta's API unfortunately does not alway provide a status on the factors. It seems that the only status that is guaranteed is the negative ("NOT_SETUP").